### PR TITLE
[IDE] Refactoring: conversion strings concatenation to interpolation …

### PR DESCRIPTION
**Pull request by Kacper20**
Initial implementation of conversion strings concatenation expression to interpolation refactor.
I would appreciate any feedback. It's my first touch to the AST & Refactoring Engine and there may be an easier way to do certain things.

I've written tests, but I'm in the middle of running it(had some problems with running through Xcode), so it may still fail. Working on it :)

Addresses [SR-5745](https://bugs.swift.org/browse/SR-5745).

cc: @nkcsgexi

### CODE REVIEW

File: lib/IDE/Refactoring.cpp [hunk 1]
```diff
@@ -1448,6 +1448,98 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
...
+  struct StringInterpolationExprFinder: public SourceEntityWalker {
+    std::unique_ptr<llvm::SetVector<Expr*>> Bucket = llvm::make_unique<llvm::SetVector<Expr*>>();
+    bool IsValidInterpolation = true;
+    StringInterpolationExprFinder() {}
+
+    bool isConcatenationExpr(DeclRefExpr* Expr) {
+      if (!Expr)
+        return false;
+      auto *FD = dyn_cast<FuncDecl>(Expr->getDecl());
+      if (FD == nullptr || FD->getName().str() != "+") {
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: We should compare decl pointer instead of its name for "+" here. Could you try to add a function like "getStringDecl()" in ASTContext and use it here?

File: lib/IDE/Refactoring.cpp [hunk 2]
```diff
@@ -1448,6 +1448,98 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
...
+    bool walkToExprPre(Expr *E) {
+      if (E->isImplicit())
+        return true;
+
+      auto ExprType = E->getType().getString();
+      //Only binary concatenation operators should exist in expression
+      if (E->getKind() == ExprKind::Binary) {
+        auto *BE = dyn_cast<BinaryExpr>(E);
+        auto *OperatorDeclRef = BE->getSemanticFn()->getMemberOperatorRef();
+        if (!isConcatenationExpr(OperatorDeclRef) || ExprType != "String") {
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: Same here. We should use `ExprType->getNominalOrBoundGenericNominal() == Ctx.getStringDecl()` here instead of comparing strings.

File: lib/IDE/Refactoring.cpp [hunk 3]
```diff
@@ -1448,6 +1448,98 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
...
+          IsValidInterpolation = false;
+          return false;
+        }
+        return true;
+      }
+      //Everything that evaluates to string should be gathered
+      if (ExprType == "String") {
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: Same here as above.

File: lib/IDE/Refactoring.cpp [hunk 4]
```diff
@@ -1448,6 +1448,98 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
...
+        Bucket->insert(E);
+        return false;
+      }
+      if (E->getKind() == ExprKind::DeclRef) {
+        auto *DR = dyn_cast<DeclRefExpr>(E);
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: The above two lines can be combined to `if (auto *DR = dyn_cast<DeclRefExpr>(E)) {`. Also, could you add a comment above why we need to handle this case specifically?

File: lib/IDE/Refactoring.cpp [hunk 5]
```diff
@@ -1448,6 +1448,98 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
...
+      //There was non-expected expression, it's not valid interpolation then
+      IsValidInterpolation = false;
+      return false;
+    }
+  } Walker;
+  Walker.walk(E);
+  //There should be two or more expressions to convert
+  if (!Walker.IsValidInterpolation || Walker.Bucket->size() < 2)
+    return nullptr;
+  return std::move(Walker.Bucket);
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: We don't need `std::move` here.

File: lib/IDE/Refactoring.cpp [hunk 6]
```diff
@@ -1448,6 +1448,98 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
...
+}
+
+StringRef static interpolatedExpressionForm(Expr* E, SourceManager& SM) {
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: `static` before `StringRef`.

File: lib/IDE/Refactoring.cpp [hunk 7]
```diff
@@ -1448,6 +1448,98 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
...
+  if (auto *Literal = dyn_cast<StringLiteralExpr>(E)) {
+    return Literal->getValue();
+  }
+  auto CharRange = Lexer::getCharSourceRangeFromSourceRange(SM, E->getSourceRange());
+  auto ExpStr = SM.extractText(CharRange).str();
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: The above two lines can be combined as 
```
auto ExpStr = Lexer::getCharSourceRangeFromSourceRange(SM, E->getSourceRange()).str();
```

**Review by nkcsgexi — CHANGES_REQUESTED** [review]
Great work so far! Some comments inline.

File: lib/IDE/Refactoring.cpp (hunk 5 above)
Comment (Kacper20) [reply lib/IDE/Refactoring.cpp]: Hm, I'm returning `std::unique_ptr` which has deleted copy constructor, so I need to move the membership if the func signature is `static std::unique_ptr<llvm::SetVector<Expr*>>
  findConcatenatedExpressions(ResolvedRangeInfo Info)`. Am I missing something?

**Comment by Kacper20** [comment]
Thanks for the review! Working on fixes right now :)

File: lib/IDE/Refactoring.cpp (hunk 5 above)
Comment (nkcsgexi) [reply lib/IDE/Refactoring.cpp]: oh, you are right. We don't need `std::move` if the unique pointer is a local variable.

Comment (Kacper20) [reply lib/IDE/Refactoring.cpp]: Great, thanks!

File: lib/IDE/Refactoring.cpp (hunk 1 above)
Comment (Kacper20) [reply lib/IDE/Refactoring.cpp]: Hmm... Could you help me a bit on this? I've changed the other string comparisons but this one is a bit more tricky. What I've done is: I've found `KnownDecls.def` which seems to define functions and is a potential place to dig in. I could add a new entry and as I see there is a macro that generated concrete implementation of the getter. But how should I name it(string literal form)? How does binding look like between this declaration and the object in memory? I haven't found any examples for using `FUNCDECL` macro with operators like `+`. Thanks in advance!

Comment (nkcsgexi) [reply lib/IDE/Refactoring.cpp]: I think we can use the function I added in #11964. Please let me know if that works.

File: lib/IDE/Refactoring.cpp [hunk 8]
```diff
@@ -1448,6 +1448,98 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
...
+      }
+      return true;
+    }
+
+    bool walkToExprPre(Expr *E) {
+      if (E->isImplicit())
+        return true;
+
+      auto ExprType = E->getType().getString();
+      //Only binary concatenation operators should exist in expression
```
Comment (xwu) [inline lib/IDE/Refactoring.cpp]: Drive-by nit: comments should have a space after `//` and end in punctuation.

File: lib/IDE/Refactoring.cpp (hunk 1 above)
Comment (Kacper20) [reply lib/IDE/Refactoring.cpp]: Thanks for this! Unfortunately `FD` and `ASTContext getPlusFunctionOnRangeReplaceableCollection()` are not the same. That's weird since String from Swift 4 adopts this protocol. I'm using `FD` like in the example above. Should I maybe refer to it in some other way?

Comment (Kacper20) [reply lib/IDE/Refactoring.cpp]: Hm, it's defined directly in `String.swift` here: https://github.com/apple/swift/blob/master/stdlib/public/core/String.swift#L873 It's probably why it doesn't work.

**Comment by Kacper20** [comment]
Hi @nkcsgexi,

I've fixed all of the code according to the comments - thanks for them! I've implemented new function on `ASTContext` to get the + function declaration for Strings. Many thanks for showing me the way, it was really helpful and I learned a lot.

Two code transformation test cases are passing now. `RefactoringKind/basic.swift` test is failing but it's not in the code I've written: `118:8: error: expected '{' in body of function declaration`. Should this be like that, or should I fix it?

Please let me know whether I can improve something.

File: lib/IDE/Refactoring.cpp (hunk 1 above)
Comment (Kacper20) [reply lib/IDE/Refactoring.cpp]: I've ended up adding new function on `ASTContext`

File: lib/AST/ASTContext.cpp [hunk 9]
```diff
@@ -571,6 +574,38 @@ FuncDecl *ASTContext::getPlusFunctionOnRangeReplaceableCollection() const {
...
+      if (ParamLists.size() != 2 || ParamLists[1]->size() != 2)
+        continue;
+      auto FirstParam = ParamLists[1]->get(0);
+      auto SecondParam = ParamLists[1]->get(1);
+      auto FirstParamType =
+      FirstParam->getInterfaceType()->getNominalOrBoundGenericNominal();
+      auto SecondParamType =
+      SecondParam->getInterfaceType()->getNominalOrBoundGenericNominal();
+      auto StringDecl = getStringDecl();
+      if (FirstParamType == StringDecl && SecondParamType == StringDecl) {
```
Comment (nkcsgexi) [inline lib/AST/ASTContext.cpp]: Nitpick: Can we simplify the above logic by introducing a local closure `isStringParam` and use the closure on both `ParamLsts[1]->get[0]` and `ParamLsts[1]->get[1]`?

File: lib/IDE/Refactoring.cpp [hunk 10]
```diff
@@ -1448,6 +1448,106 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
...
+    ASTContext& Ctx;
+
+    bool IsValidInterpolation = true;
+    StringInterpolationExprFinder(ASTContext& Ctx): Ctx(Ctx) {}
+
+    bool isConcatenationExpr(DeclRefExpr* Expr) {
+      if (!Expr)
+        return false;
+      auto *FD = dyn_cast<FuncDecl>(Expr->getDecl());
+      if (FD == nullptr || FD != Ctx.getPlusFunctionOnString()) {
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: Theoretically, both this '+' and the '+' function I added in the other commit could apply to strings. It's entirely up to type checker to decide the preferred one. Could you check both situations here?

File: lib/IDE/Refactoring.cpp [hunk 11]
```diff
@@ -1448,6 +1448,106 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
...
+  Walker.walk(E);
+
+  // There should be two or more expressions to convert.
+  if (!Walker.IsValidInterpolation || Walker.Bucket->size() < 2)
+    return nullptr;
+
+  return std::move(Walker.Bucket);
+}
+
+static std::string interpolatedExpressionForm(Expr* E, SourceManager& SM) {
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: Instead of returning `std::string`, could we take a stream operator and output to the stream instead?
e.g.
```
  llvm::SmallString<64> Buffer;
  llvm::raw_svector_ostream OS(Buffer);
  interpolatedExpressionForm(E, SM, OS); // use OS << "something"
```

**Review by nkcsgexi — CHANGES_REQUESTED** [review]
Great improvement! More comments inline.

**Comment by Kacper20** [comment]
Thanks, @nkcsgexi :)
Please let me know whether it's what you've expected.

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Review by nkcsgexi — APPROVED** [review]

**Comment by nkcsgexi** [comment]
@Kacper20 mind taking a look at these failing tests?

**Comment by Kacper20** [comment]
@nkcsgexi Yes, of course. Looking at it :)

**Comment by Kacper20** [comment]
@nkcsgexi I wonder one more thing: From where should I take the `ASTContext` in `isApplicable` call? I see that I could get it through `Info.RangeContext` and `Info.getType`. What's the preferred way to do this thing?

**Comment by Kacper20** [comment]
@nkcsgexi I've fixed `refactoring/` tests but SourceKit ones are still failing. They're failing even if I'm commenting out the whole implementation, rebuilding and running tests again. Any ideas what might be wrong?

**Comment by Kacper20** [comment]
Tests are passing for me locally right now. There was weird error with SourceKit tests regarding calling pure virtual function but it doesn't occur now for me. Could we check whether this works on CI?

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by Kacper20** [comment]
@CodaFi Could you please run again? A dumb mistake from my side done while hunting those SourceKit tests crashes ;). Thanks in advance!

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@Kacper20 I think getting it from `Info.RangeContext` makes perfect sense, something like `Info.RangeContext->getASTContext()`

**Comment by Kacper20** [comment]
@nkcsgexi Any idea why build on Linux fails?

File: include/swift/IDE/RefactoringKinds.def [hunk 12]
```diff
@@ -46,6 +46,7 @@ RANGE_REFACTORING(ExtractFunction, "Extract Method", extract.function)
 
 RANGE_REFACTORING(ExtractRepeatedExpr, "Extract Repeated Expression", extract.expr.repeated)
 
+RANGE_REFACTORING(ConvertStringsConcatenationToInterpolation, "Convert to Strings Interpolation", convert.string-concatenation.interpolation)
```
Comment (harlanhaskins) [inline include/swift/IDE/RefactoringKinds.def]: Nit: This should be `Convert to String Interpolation`

Comment (Kacper20) [reply include/swift/IDE/RefactoringKinds.def]: Thanks, fixed!

File: lib/IDE/Refactoring.cpp [hunk 13]
```diff
@@ -1555,6 +1555,116 @@ bool RefactoringActionCollapseNestedIfExpr::performChange() {
...
+
+  // There should be two or more expressions to convert.
+  if (!Walker.IsValidInterpolation || Walker.Bucket->size() < 2)
+    return nullptr;
+
+  return std::move(Walker.Bucket);
+}
+
+static void interpolatedExpressionForm(Expr* E, SourceManager& SM,
+                                              llvm::raw_svector_ostream& OS) {
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: Tiny nit: could we change the second parameter to type `llvm::raw_ostream &OS`?

Comment (Kacper20) [reply lib/IDE/Refactoring.cpp]: Done. Fixed in all other places, too. Thanks!

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

File: lib/IDE/Refactoring.cpp (hunk 13 above)
Comment (nkcsgexi) [reply lib/IDE/Refactoring.cpp]: Thanks for fixing the name convention! I mean the type of the second parameter here should be `llvm:: raw_ostream` instead of `llvm::raw_svector_ostream` to be more acceptive in the future.

Comment (Kacper20) [reply lib/IDE/Refactoring.cpp]: Oh, sorry :) Right now should be fine.

**Comment by nkcsgexi** [comment]
@swift-ci Please clean smoke test

**Comment by nkcsgexi** [comment]
@harlanhaskins any more comments?

**Comment by harlanhaskins** [comment]
Nothing more from me! Great work, @Kacper20!

**Comment by Kacper20** [comment]
Thanks for all of the help in my first contribution to Swift OS @nkcsgexi, @harlanhaskins :)
